### PR TITLE
Add while loop batching

### PR DESCRIPTION
**Pull request by tchaton**
This PR adds an example of inference as a loop for stable diffusion.

<p>Computed on RTX 3090:</p>

Type | Max Batch Size | Number of users (locust) | Average (ms)
-- | -- | -- | --
Grouped Batch | 6 | 6 | 8868
Streamed Batch (naive) | 6 | 6 | 9442
Streamed Batch (global state) | 6 | 6 | 8584

And the following benchmarks on T4.

Type | Max Batch Size | Number of users (locust) | Average (ms) | Number of requests
-- | -- | -- | -- | --
Grouped Batch | 1 | 1 | 5061 | 1353
Grouped Batch | 2 | 2 | 11393 | 1599

### CODE REVIEW

File: test.py [hunk 1]
```diff
@@ -0,0 +1,15 @@
+import base64
+from pathlib import Path
```
Comment (tchaton) [inline test.py:2]: To be removed.

File: README.md [hunk 2]
```diff
@@ -6,37 +6,46 @@ To get started, save this code snippet as `app.py` and run the below at the end
...
 import lightning as L
-import os, base64, io, ldm, torch
+import os, base64, io, torch
 from diffusion_with_autoscaler import AutoScaler, BatchText, BatchImage, Text, Image, CustomColdStartProxy
 
 PROXY_URL = "https://ulhcn-01gd3c9epmk5xj2y9a9jrrvgt8.litng-ai-03.litng.ai/api/predict"
 
+class FlashAttentionBuildConfig(L.BuildConfig):
+
+    image = "ghcr.io/gridai/lightning-stable-diffusion:v0.2"
```
Comment (lantiga) [inline README.md:19]: We should probably also host the Dockerfile in the repo, for those who want to reproduce.

File: test.py (hunk 1 above)
Comment (lantiga) [reply test.py:2]: you mean the whole file right?

File: diffusion_with_autoscaler/autoscaler.py [hunk 3]
```diff
@@ -231,49 +232,81 @@ async def send_batch(self, batch: List[Tuple[str, _BatchRequestModel]], server_u
             self._responses.update(result)
         finally:
             if server_url in self._server_status:
-                # TODO - if the server returns an error, track that so
-                #  we don't send more requests to it
-                self._server_status[server_url] = True
+                async with self._lock:
+                    if self.batching == "streamed":
+                        # decrement the number of currently processed requests.
+                        self._server_status[server_url] -= 1
```
Comment (lantiga) [inline diffusion_with_autoscaler/autoscaler.py:238]: Assert that it's not negative for sanity?

File: diffusion_with_autoscaler/autoscaler.py [hunk 4]
```diff
@@ -231,49 +232,81 @@ async def send_batch(self, batch: List[Tuple[str, _BatchRequestModel]], server_u
...
-            if batch and (is_batch_ready or is_batch_timeout):
-                self._server_status[server_url] = False
-                # find server with capacity
-                asyncio.create_task(self.send_batch(batch, server_url))
-                # resetting the batch array, TODO - not locking the array
-                self._batch = self._batch[len(batch) :]
-                self._last_batch_sent = time.time()
+        try:
+            while True:
+                await asyncio.sleep(0.05)
```
Comment (lantiga) [inline diffusion_with_autoscaler/autoscaler.py:268]: it's ok to sleep here but we don't have to, strictly speaking

Comment (lantiga) [reply diffusion_with_autoscaler/autoscaler.py:268]: but let's keep it in, maybe we can take it down to 0.01 (have you tried different values?)

**Review by lantiga — APPROVED** [review]
Looks great, minor comments

File: test.py (hunk 1 above)
Comment (tchaton) [reply test.py:2]: Yes.

File: README.md (hunk 2 above)
Comment (tchaton) [reply README.md:19]: I will add a TODO for it, but right now. It doesn't seem to work.

File: diffusion_with_autoscaler/autoscaler.py (hunk 3 above)
Comment (tchaton) [reply diffusion_with_autoscaler/autoscaler.py:238]: Done !

File: diffusion_with_autoscaler/autoscaler.py (hunk 4 above)
Comment (tchaton) [reply diffusion_with_autoscaler/autoscaler.py]: Changed to 0.001

**Pull request merged**